### ISSUE TITLE
Could the numeric precision of mxnet be less than double?

### ISSUE THREAD

**Issue by predict-r** [apache/incubator-mxnet#3088]
R uses double precision math, and I thought mxnet cpu does too (?). If so, why is the accuracy of mxnet a few percentages behind another model calculated using R's normal BLAS on a OS X operating system? 

```
# Stanford cs231n example 
# Sprial data

N = 100 # number of points per class
D = 2 # dimensionality
K = 3 # number of classes
num_examples = N*K
X = matrix(0,num_examples,D) # data matrix (each row = single example)
y = rep(0,num_examples) # class labels

for (j in 0:(K-1)){
  ix = seq(N*j+1,N*(j+1))
  r = seq(0.0,1,length.out=N) # radius
  t = seq(j*4,(j+1)*4,length.out=N) + rnorm(N)*0.2 # theta
  X[ix,1] = r*sin(t)
  X[ix,2]= r*cos(t)
  y[ix] = j +1
}
# lets visualize the data:
points(X[, 1], X[, 2], col=rainbow(3)[y])

# Neural network using standard R blas

hidden = h = 100

W = list()
b = list()
    W[[1]]= matrix(rnorm(ncol(X)*hidden, sd=0.01),D,h)
    b[[1]] = rep(0,h)
     nClasses = max(y)
    W[[2]]= matrix(rnorm(h*K, sd=0.01),h ,K  )
    b[[2]] = rep(0,K)


step_size  = 1e-0
reg = 1e-3


for(i in 1:10000){

hidden_layer = X %*% W[[1]] + matrix(b[[1]],nrow(X),ncol(W[[1]]),byrow=T) 
hidden_layer[hidden_layer<0] <-0

scores = hidden_layer %*% W[[2]] + matrix(b[[2]],nrow(hidden_layer),ncol(W[[2]]),byrow=T) 
predicted_class = apply(scores,1, which.max)

exp_scores = exp(scores)
probs = exp_scores/rowSums(exp_scores)

correct_logprobs = numeric()
for(row in 1:nrow(probs)){
  correct_logprobs =  -log(probs[row, y[row]])
}

data_loss = sum(correct_logprobs)/num_examples
reg_loss = 0.5 * reg * sum(W[[1]]*W[[1]]) + 0.5 * reg * sum(W[[2]]*W[[2]])
loss = data_loss + reg_loss

if(i %% 1000==0) {
  cat("\n iteration",i,": loss",loss," ,accuracy: ",mean(predicted_class==y))
  plot(X[, 1], X[, 2], col=rainbow(3)[y], pch = ifelse(predicted_class==y,1,4)  )

} 

dscores = probs
for(row in 1:nrow(probs)){
dscores[row,y[row]] = dscores[row,y[row]]  -1
}
dscores= dscores/num_examples

dW2 = t(hidden_layer) %*% dscores
db2 = colSums(dscores)

dhidden = dscores %*% t(W[[2]])
dhidden[hidden_layer<=0] <- 0
dW = t(X) %*% dhidden
db = colSums(dhidden)

dW2 = dW2 + reg * W[[2]]
dW = dW + reg * W[[1]]

W[[1]] = W[[1]] - step_size * dW
b[[1]] = b[[1]] - step_size *db
W[[2]] = W[[2]] - step_size * dW2
b[[2]] = b[[2]] - step_size * db2

}


predict.diy <- function(W,b,X){
  hidden_layer = X %*% W[[1]] + matrix(b[[1]],nrow(X),ncol(W[[1]]),byrow=T) 
  hidden_layer[hidden_layer<0] <-0

  scores = hidden_layer %*% W[[2]] + matrix(b[[2]],nrow(hidden_layer),ncol(W[[2]]),byrow=T) 
  predicted_class = apply(scores,1, which.max)

  exp_scores = exp(scores)
  probs = exp_scores/rowSums(exp_scores)
  return(probs)
}

predicted_class = apply(predict.diy(W,b,X),1, which.max)
cat("\naccuracy",mean(predicted_class==y))
# accuracy is 99% after 3000 iterations and at 10,000

# MXNET
model = mx.mlp(X, y-1, 
               hidden_node = c(100), 
               out_node = 3, 
               num.round = 10000,
               out_activation = "softmax",
               learning.rate = 1e-0,
               wd = 1e-3,
               array.layout = "rowmajor",
               initializer=mx.init.normal(0.01))


predicted_class = apply(as.array(predict(model, X, array.layout = 'rowmajor')),2,which.max)
cat("accuracy:",mean(predicted_class==y),"\n")
plot(X[, 1], X[, 2], col=rainbow(3)[y], pch = ifelse(predicted_class==y,1,4)  )
# accuracy is 95% after 10,000 iterations


```

**Comment by piiswrong** [comment]
mxnet uses float by default

**Comment by predict-r** [comment]
Thank you that answers my question. Is there a setting to use different precisions, like in Torch? I can imagine using low precision most of the time and fine tuning with high precision.

**Comment by ufukcbicici** [comment]
I also want to ask the same question. I am using the python library and have built a new algorithm and I want to make sure that my gradients are correct. I am applying finite differences method for that and as usual, this is highly dependent on the fact that we are having reasonable floating point precision. Gpu uses f32 as usual. But is there any way to chance the precision to double, at least for the Cpu ?

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
